thunder/core/transforms.py

```diff
@@ -2522,6 +2522,8 @@ def is_constant_for_vjp(symbol: prims.Symbol) -> bool:
     Returns:
         bool: True if the symbol is constant, False otherwise.
     """
+    if isinstance(symbol.sym.id, str) and symbol.sym.id.startswith("higher_order_autograd_function_apply"):
+        return False
```

Comment:
Why is this needed? Can you please provide an example of a failing case without these lines?

Reply:
the reason is https://github.com/Lightning-AI/lightning-thunder/pull/1548#issuecomment-2551343344, the test case `test_splitter_autograd_function` modified in this PR will fail without it

Reply:
It's most likely a viable workaround for the problems seen in this PR, but I don't think this is a correct fix. There's a problem with `thunder.torch._set_grad_enabled_with_warning(False)` called inside the forward function passed to `thunder.torch.autograd_function_apply` that causes the system to ignore provided backward which should be fixed first:
```py
import thunder
import torch

def forward(_, x):
    saved_for_backward = (x,)
    thunder.torch._set_grad_enabled_with_warning(False) # Without this line the specified backward is called as expected
    sin = thunder.torch.sin(x)
    thunder.torch._set_grad_enabled_with_warning(True)
    return sin, saved_for_backward

def backward(_, grad_output, *saved_tensors):
    raise NotImplementedError

def my_sin(x):
    res = thunder.torch.autograd_function_apply(
        forward,
        backward,
        x,
        args_tensor_mask=[True],
        non_differentiable_idx=[],
    )
    return res

jitted = thunder.jit(my_sin)
x = torch.randn((2, 2), requires_grad=True)

out = jitted(x) # Should raise NotImplementedError but it doesn't
out.backward(out)
```

Reply:
yes, for the thunderFX path it always adds a no_grad guard around the forward function, so it's no problem. But in the thunder.jit path we need to think of how to handle the situation that there's no_grad in the forward function

Reply:
Thanks for the repro, I think `autograd_function_apply` should override any `set_grad_enabled` inside. I will try updating #1463 to work with this and discuss the fix.

Reply:
`thunder.torch.autograd_function_apply` corresponds to the `torch.ops.higher_order.autograd_function_apply` that only appears in dynamo, and dynamo will add a no_grad guard around the forward function, so it's ok if we use it in thunderFX, but I agree we need to think if there are other ways to handle it. 
if we write it as follows, it raises error
```
import thunder
import torch
from thunder.dynamo import thunderfx

class Sin(torch.autograd.Function):
    @staticmethod
    def forward(ctx, x):
        ctx.save_for_backward(x)
        with torch.no_grad():
            return torch.sin(x)

    @staticmethod
    def backward(ctx, g):
        #(x,) = ctx.saved_tensors
        #return g * torch.cos(x) * 100
        raise NotImplementedError("aaaaa")

def my_sin(x):
    return Sin.apply(x)

# jitted = thunder.jit(my_sin)
jitted = thunderfx(my_sin)
x = torch.randn((2, 2), requires_grad=True)

out = jitted(x) # Should raise NotImplementedError but it doesn't
out.backward(out)
```

Reply:
I think it works for the above repro with `thunderfx` because `torch.compile` leads to graph-break when it sees an exception, so we fallback to eager and eager mode throws the error.

Running the above script with `TORCH_LOGS="graph_breaks"` shows

```python
V1219 11:49:04.457000 2947898 torch/_dynamo/convert_frame.py:1263] [0/0] [__graph_breaks]   File "/home/kkalambarkar/git/pytorch/torch/_dynamo/symbolic_convert.py", line 2914, in run
V1219 11:49:04.457000 2947898 torch/_dynamo/convert_frame.py:1263] [0/0] [__graph_breaks]     super().run()
V1219 11:49:04.457000 2947898 torch/_dynamo/convert_frame.py:1263] [0/0] [__graph_breaks]   File "/home/kkalambarkar/git/pytorch/torch/_dynamo/symbolic_convert.py", line 1120, in run
V1219 11:49:04.457000 2947898 torch/_dynamo/convert_frame.py:1263] [0/0] [__graph_breaks]     while self.step():
V1219 11:49:04.457000 2947898 torch/_dynamo/convert_frame.py:1263] [0/0] [__graph_breaks]   File "/home/kkalambarkar/git/pytorch/torch/_dynamo/symbolic_convert.py", line 1035, in step
V1219 11:49:04.457000 2947898 torch/_dynamo/convert_frame.py:1263] [0/0] [__graph_breaks]     self.exception_handler(e)
V1219 11:49:04.457000 2947898 torch/_dynamo/convert_frame.py:1263] [0/0] [__graph_breaks]   File "/home/kkalambarkar/git/pytorch/torch/_dynamo/symbolic_convert.py", line 1631, in exception_handler
V1219 11:49:04.457000 2947898 torch/_dynamo/convert_frame.py:1263] [0/0] [__graph_breaks]     raise Unsupported("Observed exception")
V1219 11:49:04.457000 2947898 torch/_dynamo/convert_frame.py:1263] [0/0] [__graph_breaks] torch._dynamo.exc.Unsupported: Observed exception
V1219 11:49:04.457000 2947898 torch/_dynamo/convert_frame.py:1263] [0/0] [__graph_breaks] 
V1219 11:49:04.457000 2947898 torch/_dynamo/convert_frame.py:1263] [0/0] [__graph_breaks] from user code:
V1219 11:49:04.457000 2947898 torch/_dynamo/convert_frame.py:1263] [0/0] [__graph_breaks]    File "/home/kkalambarkar/lightning-thunder/scratchpad/test.py", line 267, in my_sin
V1219 11:49:04.457000 2947898 torch/_dynamo/convert_frame.py:1263] [0/0] [__graph_breaks]     return torch.cos(x) + Sin.apply(x) # + torch.cos(x)
```

Regarding the fix, this is what I have in mind (though I will think a bit more about this)
```patch
diff --git a/thunder/torch/__init__.py b/thunder/torch/__init__.py
index 38d0353c..e59a66e8 100644
--- a/thunder/torch/__init__.py
+++ b/thunder/torch/__init__.py
@@ -5655,6 +5655,25 @@ else:
         utils.check(False, lambda: "torch.distributed is not available")
 
 
+def call_higher_order_function_and_consider_outer_autograd_setting(fn):
+    # In case of higher order function like `autograd_function_apply`,
+    # whether the output should be tagged with `DETACHED_AUTOGRAD_GRAPH`
+    # depends on whether `autograd_function_apply` was called with grad enabled or not.
+    # It shouldn't depend on whether grad was enabled or disabled inside the `fwd` (or the function)
+    # called by the higher order operator.
+
+    @wraps(fn)
+    def wrapper(*args, **kwargs):
+        is_grad_enabled = get_compile_data().is_grad_enabled
+
+        result, saved_for_backward = fn(*args, **kwargs)
+        if is_grad_enabled and ProxyTag.DETACHED_AUTOGRAD_GRAPH in result.tags:
+            result.tags.remove(ProxyTag.DETACHED_AUTOGRAD_GRAPH)
+        return result, saved_for_backward
+
+    return wrapper
+
+
 # ref: https://github.com/pytorch/pytorch/blob/b99ef1a/torch/_functorch/autograd_function.py#L715-L752
 @torchsymbol(
     torch.ops.higher_order.autograd_function_apply,
@@ -5668,7 +5687,7 @@ def autograd_function_apply(
     args_tensor_mask: Sequence[bool] | None,
     non_differentiable_idx: Sequence[int] | None = None,
 ) -> TensorProxy | tuple[TensorProxy, ...]:
-    result, saved_for_backward = fwd(None, *args)
+    result, saved_for_backward = call_higher_order_function_and_consider_outer_autograd_setting(fwd)(None, *args)
     return result
```

Reply:
> I think it works for the above repro with thunderfx because torch.compile leads to graph-break when it sees an exception, so we fallback to eager and eager mode throws the error.

it seems dynamo has some checks for raising error in backward function, it doesn't goes to thunder. If you change the backward function to 
```
@staticmethod
    def backward(ctx, g):
        (x,) = ctx.saved_tensors
        return g * torch.cos(x) * 100
        #raise NotImplementedError("aaaaa")
```
thunderFX can give the right bwd trace

Reply:
@kshitij12345, could you please create a pull request with the changes from https://github.com/Lightning-AI/lightning-thunder/pull/1548#discussion_r1891614613 and a test from https://github.com/Lightning-AI/lightning-thunder/pull/1548#discussion_r1891386370. With your fix, the code added here to `is_constant_for_vjp` shouldn't be necessary.

Reply:
Have updated #1463

Reply:
Thank you @kshitij12345 , I've removed the cherry-picked commits and the modification in `is_constant_for_vjp`, I think after https://github.com/Lightning-AI/lightning-thunder/pull/1548#discussion_r1891960630 is fixed, the test case `test_splitter_autograd_function` in this PR should pass